### PR TITLE
Adds USB comms to the Tessel 2

### PR DESCRIPTION
**Pull request by johnnyman727**

### CODE REVIEW

**Comment by johnnyman727** [comment]
This commit is functional in conjunction with the following firmware commit: https://github.com/tessel/v2-firmware/commit/92f746e7834d0f13d6d66ecb6de5be2b7be806b1

**Comment by johnnyman727** [comment]
This commit is also dependant on this PR in the openwrt image: https://github.com/tessel/openwrt-tessel/pull/6

File: lib/lan_connection.js [hunk 1]
```diff
@@ -64,6 +73,18 @@ LANConnection.prototype.open = function() {
   });
 }
 
+LANConnection.prototype._processArgsForTransport = function(command) {
+
+  // Ensure the args are an array
+  if (!Array.isArray(command)) return;
+
+  // Join the args into a string
+  command = command.join(' ');
```
Comment (kevinmehall) [inline lib/lan_connection.js]: this should involve escaping

File: lib/usb_connection.js [hunk 2]
```diff
@@ -70,17 +102,74 @@ USBConnection.prototype.open = function(callback) {
...
+USBConnection.prototype._processArgsForTransport = function(command) {
+
+  if (!Array.isArray(command)) return;
+
+  function quoteRemover(match, p1, p2, p3) {
+    return p2;
+  }
+
+  // For each command
+  command.forEach(function(arg, i) {
```
Comment (kevinmehall) [inline lib/usb_connection.js]: Things passed to this shouldn't be quoted. Escape strings for SSH transport, don't unescape them here. It only needs to check that the array items don't themselves contain null bytes, but can probably just reject those.

**Comment by johnnyman727** [comment]
@HarleyKwyn If you feel like refactoring, now is a good time :)

**Comment by johnnyman727** [comment]
I can also do it tomorrow

**Comment by HarleyKwyn** [comment]
Don't work on the weekends :P Go outside and enjoy the day.

**Comment by johnnyman727** [comment]
This is blocked until #63 is fixed.

**Comment by johnnyman727** [comment]
Should be reviewed by @tcr.

File: lib/tessel/commands.js [hunk 3]
```diff
@@ -1,30 +1,42 @@
...
 }
-module.exports.runOnBoot = function(filepath) {
-  return 'echo "#!/bin/sh\ncd /app\nexec node ." > ' + filepath + '/start && chmod +x "' + filepath + '/start"';
+module.exports.openStdinToFile = function(filepath) {
+  return ['dd', 'of=' + filepath];
 }
-module.exports.startPushedScript = function() {
-  return '/etc/init.d/tessel-app start';
+module.exports.setExecutablePermissions = function(filepath) {
+  return ['chmod', 'x', filepath];
```
Comment (jiahuang) [inline lib/tessel/commands.js]: does this need to be +x or does x work?

File: lib/tessel/deploy.js [hunk 4]
```diff
@@ -18,76 +20,141 @@ Tessel.prototype.deployScript = function(opts, push) {
...
+function pushScript(t, filepath) {
+  // Write the node start file
+  writeToFile(t, startPushedScript.bind(this, t));
+}
+
+function writeToFile(t, callback) {
+  t.connection.exec(commands.openStdinToFile(PUSH_PATH + PUSH_START_SCRIPT_NAME), function(err, remoteProcess) {
+    if (err) throw err;
+    remoteProcess.stdin.once('finish', function() {
+      tessel.logs.info('You may now disconnect from the Tessel. Your code will be run whenever Tessel boots up. To remove this code, use `tessel erase`.');
```
Comment (jiahuang) [inline lib/tessel/deploy.js]: is the "Your code will run whenever Tessel boots up" part true right now?

Comment (johnnyman727) [inline lib/tessel/deploy.js]: Yeah

File: lib/tessel/commands.js (hunk 3 above)
Comment (johnnyman727) [inline lib/tessel/commands.js]: You're right - good catch!

**Comment by johnnyman727** [comment]
@Frijol caught an issue where the start script didn't have the permissions set properly when pushed. I'll need to fix that prior to merging.

**Comment by Frijol** [comment]
Other than the permissions issue @johnnyman727 notes, every function listed in the Readme works as listed

**Comment by johnnyman727** [comment]
Okay, this PR should be ready to merge.

**Comment by Frijol** [comment]
lgtm

**Pull request merged**